File: Chroma-Core/src/main/java/buttondevteam/lib/player/TBMCPlayerBase.java
package buttondevteam.lib.player;

import buttondevteam.lib.architecture.ConfigData;
import buttondevteam.lib.architecture.IHaveConfig;
import lombok.Getter;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;

import java.util.Set;
import java.util.UUID;

@AbstractUserClass(foldername = "minecraft", prototype = TBMCPlayer.class)
@TBMCPlayerEnforcer
public abstract class TBMCPlayerBase extends ChromaGamerBase {
	protected UUID uuid;

	@Getter
	private final IHaveConfig config = new IHaveConfig(this::save);

	public UUID getUUID() {
		if (uuid == null)
			uuid = UUID.fromString(getFileName());
		return uuid;
	}

	public ConfigData<String> PlayerName() {
		return super.config.getData("PlayerName", "");
	}

	/**
	 * Get player as a plugin player.
	 *
	 * @param uuid The UUID of the player to get
	 * @param cl   The type of the player
	 * @return The requested player object
	 */
	public static <T extends TBMCPlayerBase> T getPlayer(UUID uuid, Class<T> cl) {
		var player = ChromaGamerBase.getUser(uuid.toString(), cl);
		if (!player.getUUID().equals(uuid)) //It will be set from the filename because we check it for scheduling the uncache.
			throw new IllegalStateException("Player UUID differs after converting from and to string...");
		return player;
	}

	@Override
	public void init() {
		super.init();

		String pluginname;
		if (getClass().isAnnotationPresent(PlayerClass.class))
			pluginname = getClass().getAnnotation(PlayerClass.class).pluginname();
		else
			throw new RuntimeException("Class not defined as player class! Use @PlayerClass");

		var playerData = commonUserData.getPlayerData();
		var section = playerData.getConfigurationSection(pluginname);
		if (section == null) section = playerData.createSection(pluginname);
		config.reset(section);
	}

	@Override
	protected void scheduleUncache() { //Don't schedule it, it will happen on quit - if the player is online
		var p = Bukkit.getPlayer(getUUID());
		if (p == null || !p.isOnline())
			super.scheduleUncache();
	}

	/**
	 * This method returns a TBMC player from their name. See {@link Bukkit#getOfflinePlayer(String)}.
	 *
	 * @param name The player's name
	 * @return The {@link TBMCPlayer} object for the player
	 */
	@SuppressWarnings("deprecation")
	public static <T extends TBMCPlayerBase> T getFromName(String name, Class<T> cl) {
		OfflinePlayer p = Bukkit.getOfflinePlayer(name);
		return getPlayer(p.getUniqueId(), cl);
	}

	@Override
	protected void save() {
		Set<String> keys = commonUserData.getPlayerData().getKeys(false);
		if (keys.size() > 1) // PlayerName is always saved, but we don't need a file for just that
			super.save();
	}
}

File: Chroma-Core/src/main/java/buttondevteam/lib/player/TBMCPlayerBase.kt
package buttondevteam.lib.player

import buttondevteam.lib.architecture.IHaveConfig
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.util.*

@AbstractUserClass(foldername = "minecraft", prototype = TBMCPlayer::class)
@TBMCPlayerEnforcer
abstract class TBMCPlayerBase : ChromaGamerBase() {
    val uniqueId: UUID by lazy { UUID.fromString(fileName) }
    val player by lazy { Bukkit.getPlayer(uniqueId) }
    val offlinePlayer by lazy { Bukkit.getOfflinePlayer(uniqueId) }

    /**
     * If the player is the console
     */
    val isConsole = uniqueId == UUID(0, 0)

    @JvmField
    val playerName = super.config.getData("PlayerName", "")
    public override fun initConfig() {
        super.initConfig()
        val pluginName = if (javaClass.isAnnotationPresent(PlayerClass::class.java))
            javaClass.getAnnotation(PlayerClass::class.java).pluginname
        else
            throw RuntimeException("Class not defined as player class! Use @PlayerClass")
        val playerData = commonUserData.playerData
        val section = playerData.getConfigurationSection(pluginName) ?: playerData.createSection(pluginName)
        config = IHaveConfig({ save() }, section)
    }

    /**
     * If the player is online, we don't uncache, it will be uncached when they log out
     */
    override fun scheduleUncache() {
        val p = Bukkit.getPlayer(uniqueId)
        if (p == null || !p.isOnline) super.scheduleUncache()
    }

    override fun save() {
        val keys = commonUserData.playerData.getKeys(false)
        if (keys.size > 1) // PlayerName is always saved, but we don't need a file for just that
            super.save()
    }

    override fun sendMessage(message: String) {
        // TODO: Random senders (Discord) won't receive messages. Including when trying to chat.
        player?.sendMessage(message)
    }

    override fun sendMessage(message: Array<out String>) {
        player?.sendMessage(*message)
    }

    override fun getName(): String {
        return playerName.get()
    }

    companion object {
        /**
         * Get player as a plugin player.
         *
         * @param uuid The UUID of the player to get
         * @param cl   The type of the player
         * @return The requested player object
         */
        @JvmStatic
        fun <T : TBMCPlayerBase> getPlayer(uuid: UUID, cl: Class<T>): T {
            val player = getUser(uuid.toString(), cl)
            check(player.uniqueId == uuid) { "Player UUID differs after converting from and to string..." }
            return player
        }

        /**
         * This method returns a TBMC player from their name. See [Bukkit.getOfflinePlayer].
         *
         * @param name The player's name
         * @return The [TBMCPlayer] object for the player
         */
        @Suppress("deprecation")
        @JvmStatic
        fun <T : TBMCPlayerBase> getFromName(name: String, cl: Class<T>): T {
            val p = Bukkit.getOfflinePlayer(name)
            return getPlayer(p.uniqueId, cl)
        }

        fun <T : TBMCPlayerBase> getConsole(cl: Class<T>): T {
            return getPlayer(UUID(0, 0), cl)
        }

        fun getConsole(): TBMCPlayerBase {
            return getConsole(TBMCPlayerBase::class.java)
        }

        fun Player.asTBMC(): TBMCPlayerBase {
            return getPlayer(uniqueId, TBMCPlayerBase::class.java)
        }
    }
}